Refresher.Core/Pipelines/Pipeline.cs

```diff
@@ -0,0 +1,100 @@
+using System.Collections.Frozen;
+using System.Diagnostics;
+
+using GlobalState = Refresher.Core.State;
+
+namespace Refresher.Core.Pipelines;
+
+public abstract class Pipeline
+{
+    public abstract string Id { get; }
+    public abstract string Name { get; }
+    
+    public Dictionary<string, string> Inputs = [];
+    public FrozenSet<StepInput> RequiredInputs { get; private set; }
+    
+    public PipelineState State { get; private set; } = PipelineState.NotStarted;
+    
+    public float Progress
+    {
+        get
+        {
+            float completed = (this._currentStepIndex - 1) / (float)this._steps.Count;
+            float currentStep = this._currentStep?.Progress ?? 0f;
+            float stepWeight = 1f / this._steps.Count;
+            
+            return completed + currentStep * stepWeight;
+        }
+    }
+
+    public float CurrentProgress => this._currentStep?.Progress ?? 0;
+
+    protected abstract List<Type> StepTypes { get; }
+    private List<Step> _steps = [];
+    
+    private byte _currentStepIndex;
```

Comment:
I don't see a reason for this to be a `byte`, an `int` is a much more natural choice for an index in C#, and an `int` will perform better, since the JIT has to do a `MOVZX` to zero-extend the `byte` -> `int` before it can convert it to a floating point number, since theres no direct instruction to convert from `byte` -> `float` in the x86 instruction set.

See the following functions with their matching assembly.

`int`:
```c#
    static float Square(int num) {
        return (float)num; 
    }
```
```asm
       vxorps   xmm0, xmm0, xmm0
       vcvtsi2ss xmm0, xmm0, edi
       ret      
```

`byte`:
```c#
    static float Square(byte num) {
        return (float)num; 
    }
```
```asm
       movzx    rax, dil
       vxorps   xmm0, xmm0, xmm0
       vcvtsi2ss xmm0, xmm0, eax
       ret     
```

While the performance hit is going to be completely un-noticable and does not matter in the slightest, and it's unlikely for a pipeline to have more than 255 steps, I feel its an arbitrary limitation to use a byte here, this should be an `int`.